# Pandas - Basics

When wrangling data with Python, we always begin by importing Python packages like pandas so we could access its tools (or methods) which are useful for data manipulation.

**pandas** is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language.

*Description from https://pandas.pydata.org/ 

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Pandas_logo.svg/1200px-Pandas_logo.svg.png" width=500/>


## Importing the module

In [2]:
# Import library
import pandas as pd

Note that we use **as** to alias the library. We do this so we won't have to call the whole name of the module in our code.

For example, if I want to create an empty dataframe without alias, I have to call pandas.DataFrame. Meanwhile, if I use **import pandas as pd**, I only need to call pd.DataFrame.

## Loading the data

*Data from [Global Findex Database 2017](https://microdata.worldbank.org/index.php/catalog/3324#metadata-identification).*

The most common source of simple data is csv or comma-separated values file. Pandas can read other file types as well such as:
- read_excel
- read_sql
- read_json

**In terms of reading files**, we usually pass the path of the file we want pandas to read. In this case, since micro_world.csv exists inside the same folder as our notebook, we only need to indicate the file name. However, if it's inside a folder like **data**, we need to pass **'data/micro_world.csv'**.

In [3]:
# Read the data
data = pd.read_csv('micro_world.csv', engine='python')

Pandas tries to determine what dtype to set by analyzing the data in each column. A ```DtypeWarning``` is raised when the dataset read has different dtypes in a column from a file. Recall that there are different dtypes.

We have access to numpy dtypes: `float`, `int`, `bool`, `timedelta64[ns]` and `datetime64[ns]`. Note that the numpy date/time dtypes are not time zone aware. Pandas extends this set of dtypes with its own: `datetime64[ns, <tz>]`, `category`, `Int64`, `string`, `boolean`, etc.

Read the complete reference here: [Pandas dtype reference](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html).

## Exploring the data 

In [5]:
# Show the data
data

economy economycode                                    regionwb  \
0       Afghanistan         AFG                                  South Asia   
1       Afghanistan         AFG                                  South Asia   
2       Afghanistan         AFG                                  South Asia   
3       Afghanistan         AFG                                  South Asia   
4       Afghanistan         AFG                                  South Asia   
...             ...         ...                                         ...   
154918     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154919     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154920     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154921     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154922     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   

         pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
0       19456060.0    139880545  1.936754       1  18.0   2.0      1  ...   
1       19456060.0    111705622  0.332867       2  25.0   1.0      5  ...   
2       19456060.0    128866217  1.338906       2  40.0   1.0      2  ...   
3       19456060.0    138171768  0.275619       2  45.0   1.0      1  ...   
4       19456060.0    125156190  0.629304       1  30.0   1.0      4  ...   
...            ...          ...       ...     ...   ...   ...    ...  ...   
154918   9467437.0    167986916  0.413192       2  56.0   1.0      3  ...   
154919   9467437.0    121992603  0.576707       1  56.0   1.0      4  ...   
154920   9467437.0    137769624  1.481972       1  36.0   2.0      4  ...   
154921   9467437.0    181215643  0.259968       1  35.0   3.0      5  ...   
154922   9467437.0    133939668  0.791151       1  46.0   2.0      1  ...   

        receive_wages  receive_transfers  receive_pension  \
0                   4                  4                4   
1                   4                  4                4   
2                   4                  4                4   
3                   4                  4                4   
4                   4                  4                4   
...               ...                ...              ...   
154918              4                  4                4   
154919              4                  4                4   
154920              4                  4                4   
154921              1                  4                4   
154922              4                  3                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
0                       2.0              2          5.0        0.0   
1                       4.0              4          5.0        0.0   
2                       4.0              2          5.0        0.0   
3                       4.0              4          5.0        0.0   
4                       4.0              4          5.0        0.0   
...                     ...            ...          ...        ...   
154918                  4.0              4          5.0        0.0   
154919                  4.0              4          2.0        0.0   
154920                  4.0              4          1.0        0.0   
154921                  3.0              1          1.0        0.0   
154922                  4.0              4          3.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
0                       NaN       0.0                 NaN  
1                       NaN       0.0                 NaN  
2                       NaN       0.0                 NaN  
3                       NaN       0.0                 NaN  
4                       NaN       0.0                 NaN  
...                     ...       ...                 ...  
154918                  NaN       0.0                 NaN  
154919                  NaN       0.0                 NaN  
154920                  

To show the data, you may opt to use **print(data)**, **display(data)** or just plain **data**.

In [6]:
# Display shape
data.shape

(154923, 105)

It means that the data has 154,923 rows and 105 columns. That's a lot!

In [ ]:
# Display head
data.head()

economy economycode    regionwb   pop_adult  wpid_random       wgt  \
0  Afghanistan         AFG  South Asia  19456060.0    139880545  1.936754   
1  Afghanistan         AFG  South Asia  19456060.0    111705622  0.332867   
2  Afghanistan         AFG  South Asia  19456060.0    128866217  1.338906   
3  Afghanistan         AFG  South Asia  19456060.0    138171768  0.275619   
4  Afghanistan         AFG  South Asia  19456060.0    125156190  0.629304   

   female   age  educ  inc_q  ...  receive_wages  receive_transfers  \
0       1  18.0   2.0      1  ...            4.0                4.0   
1       2  25.0   1.0      5  ...            4.0                4.0   
2       2  40.0   1.0      2  ...            4.0                4.0   
3       2  45.0   1.0      1  ...            4.0                4.0   
4       1  30.0   1.0      4  ...            4.0                4.0   

   receive_pension  receive_agriculture  pay_utilities  remittances  \
0              4.0                  2.0            2.0          5.0   
1              4.0                  4.0            4.0          5.0   
2              4.0                  4.0            2.0          5.0   
3              4.0                  4.0            4.0          5.0   
4              4.0                  4.0            4.0          5.0   

   pay_onlne  pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
0        0.0                  NaN       0.0                 NaN  
1        0.0                  NaN       0.0                 NaN  
2        0.0                  NaN       0.0                 NaN  
3        0.0                  NaN       0.0                 NaN  
4        0.0                  NaN       0.0                 NaN  

[5 rows x 105 columns]

By default, .head() displays the first 5 rows of the dataframe. Note that indexing starts at 0. 😩

In [ ]:
# Display tail
data.tail()

economy economycode  \
29777  Colombia         COL   
29778  Colombia         COL   
29779  Colombia         COL   
29780  Colombia         COL   
29781  Colombia         COL   

                                                regionwb   pop_adult  \
29777  Latin America & Caribbean (excluding high income)  37031184.0   
29778  Latin America & Caribbean (excluding high income)  37031184.0   
29779  Latin America & Caribbean (excluding high income)  37031184.0   
29780  Latin America & Caribbean (excluding high income)  37031184.0   
29781  Latin America & Caribbean (excluding high income)  37031184.0   

       wpid_random       wgt  female   age  educ  inc_q  ...  receive_wages  \
29777    114558767  0.622617       1  40.0   2.0      5  ...            1.0   
29778    199772408  0.822416       2  52.0   1.0      1  ...            4.0   
29779    157076241  0.655882       2  62.0   1.0      3  ...            4.0   
29780    134449750  0.871887       1  17.0   2.0      4  ...            4.0   
29781    114486111  0.696229       2  57.0   2.0      2  ...            NaN   

       receive_transfers  receive_pension  receive_agriculture  pay_utilities  \
29777                4.0              4.0                  4.0            2.0   
29778                4.0              4.0                  4.0            2.0   
29779                4.0              4.0                  4.0            1.0   
29780                3.0              4.0                  4.0            1.0   
29781                NaN              NaN                  NaN            NaN   

       remittances  pay_onlne  pay_onlne_mobintbuy  pay_cash  \
29777          5.0        0.0                  0.0       1.0   
29778          2.0        0.0                  NaN       0.0   
29779          1.0        0.0                  NaN       0.0   
29780          5.0        0.0                  0.0       1.0   
29781          NaN        NaN                  NaN       NaN   

       pay_cash_mobintbuy  
29777                 1.0  
29778                 NaN  
29779                 NaN  
29780                 1.0  
29781                 NaN  

[5 rows x 105 columns]

Meanwhile, .tail() displays the last 5 rows.

In [ ]:
# Display top 20 rows
data.head(20)

economy economycode    regionwb   pop_adult  wpid_random       wgt  \
0   Afghanistan         AFG  South Asia  19456060.0    139880545  1.936754   
1   Afghanistan         AFG  South Asia  19456060.0    111705622  0.332867   
2   Afghanistan         AFG  South Asia  19456060.0    128866217  1.338906   
3   Afghanistan         AFG  South Asia  19456060.0    138171768  0.275619   
4   Afghanistan         AFG  South Asia  19456060.0    125156190  0.629304   
5   Afghanistan         AFG  South Asia  19456060.0    209050144  0.773167   
6   Afghanistan         AFG  South Asia  19456060.0    167258084  0.481786   
7   Afghanistan         AFG  South Asia  19456060.0    147754514  0.564638   
8   Afghanistan         AFG  South Asia  19456060.0    171367518  1.510448   
9   Afghanistan         AFG  South Asia  19456060.0    145691120  0.675545   
10  Afghanistan         AFG  South Asia  19456060.0    160492339  0.280141   
11  Afghanistan         AFG  South Asia  19456060.0    197893595  0.516904   
12  Afghanistan         AFG  South Asia  19456060.0    129005012  0.388837   
13  Afghanistan         AFG  South Asia  19456060.0    185526924  1.665766   
14  Afghanistan         AFG  South Asia  19456060.0    147434413  0.941063   
15  Afghanistan         AFG  South Asia  19456060.0    151537276  0.731453   
16  Afghanistan         AFG  South Asia  19456060.0    117012927  0.264683   
17  Afghanistan         AFG  South Asia  19456060.0    154497258  0.450706   
18  Afghanistan         AFG  South Asia  19456060.0    111819382  0.604836   
19  Afghanistan         AFG  South Asia  19456060.0    122025729  1.387100   

    female   age  educ  inc_q  ...  receive_wages  receive_transfers  \
0        1  18.0   2.0      1  ...            4.0                4.0   
1        2  25.0   1.0      5  ...            4.0                4.0   
2        2  40.0   1.0      2  ...            4.0                4.0   
3        2  45.0   1.0      1  ...            4.0                4.0   
4        1  30.0   1.0      4  ...            4.0                4.0   
5        1  53.0   2.0      1  ...            2.0                4.0   
6        1  36.0   2.0      2  ...            4.0                2.0   
7        1  29.0   2.0      4  ...            2.0                4.0   
8        1  21.0   1.0      5  ...            4.0                4.0   
9        1  75.0   1.0      5  ...            4.0                4.0   
10       1  25.0   1.0      5  ...            4.0                4.0   
11       2  24.0   1.0      5  ...            4.0                4.0   
12       2  31.0   1.0      5  ...            4.0                4.0   
13       1  22.0   1.0      5  ...            2.0                4.0   
14       1  25.0   2.0      2  ...            4.0                4.0   
15       2  30.0   1.0      1  ...            4.0                4.0   
16       2  35.0   2.0      5  ...            4.0                4.0   
17       2  30.0   1.0      5  ...            4.0                4.0   
18       2  24.0   1.0      1  ...            4.0                4.0   
19       2  19.0   2.0      2  ...            4.0                4.0   

    receive_pension  receive_agriculture  pay_utilities  remittances  \
0               4.0                  2.0            2.0          5.0   
1               4.0                  4.0            4.0          5.0   
2               4.0                  4.0            2.0          5.0   
3               4.0                  4.0            4.0          5.0   
4               4.0                  4.0            4.0          5.0   
5               4.0                  2.0            4.0          5.0   
6               4.0                  2.0            4.0          5.0   
7               4.0                  4.0            2.0          1.0   
8               4.0                  2.0            4.0          5.0   
9               4.0                  2.0            4.0          5.0   
10              4.0                  2.0            2.0         

You may also opt to display the top n rows by doing .head(n) where n is the number of rows you want to display. Convenient, right?

In [ ]:
# Display data info
data.iloc[:,0:20].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29782 entries, 0 to 29781
Data columns (total 20 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   economy      29782 non-null  object 
 1   economycode  29782 non-null  object 
 2   regionwb     29782 non-null  object 
 3   pop_adult    29782 non-null  float64
 4   wpid_random  29782 non-null  int64  
 5   wgt          29782 non-null  float64
 6   female       29782 non-null  int64  
 7   age          29684 non-null  float64
 8   educ         29772 non-null  float64
 9   inc_q        29782 non-null  int64  
 10  emp_in       29782 non-null  int64  
 11  fin2         29782 non-null  int64  
 12  fin3         12850 non-null  float64
 13  fin4         12053 non-null  float64
 14  fin5         15019 non-null  float64
 15  fin6         15019 non-null  float64
 16  fin7         29782 non-null  int64  
 17  fin8         5501 non-null   float64
 18  fin9         15507 non-null  float64
 19  fin1

The .info() displays the columns, the number of non-null records per column and the corresponding data type.

In [ ]:
# Display summary statistics regarding the data
data.iloc[:,0:20].describe()

pop_adult   wpid_random           wgt        female           age  \
count  2.978200e+04  2.978200e+04  29782.000000  29782.000000  29684.000000   
mean   1.567220e+08  1.611280e+08      1.000205      1.538379     42.775266   
std    3.656972e+08  2.879232e+07      0.691870      0.498533     18.054462   
min    1.135982e+06  1.111113e+08      0.191998      1.000000     15.000000   
25%    6.211794e+06  1.362978e+08      0.467294      1.000000     28.000000   
50%    1.018327e+07  1.611034e+08      0.807552      2.000000     40.000000   
75%    3.049117e+07  1.860099e+08      1.324539      2.000000     57.000000   
max    1.134634e+09  2.111097e+08      4.010008      2.000000     99.000000   

               educ         inc_q        emp_in          fin2          fin3  \
count  29772.000000  29782.000000  29782.000000  29782.000000  12850.000000   
mean       1.765451      3.150494      0.627392      1.584111      1.072918   
std        0.722511      1.427889      0.483507      0.528074      0.307971   
min        1.000000      1.000000      0.000000      1.000000      1.000000   
25%        1.000000      2.000000      0.000000      1.000000      1.000000   
50%        2.000000      3.000000      1.000000      2.000000      1.000000   
75%        2.000000      4.000000      1.000000      2.000000      1.000000   
max        5.000000      5.000000      1.000000      4.000000      4.000000   

               fin4          fin5          fin6          fin7         fin8  \
count  12053.000000  15019.000000  15019.000000  29782.000000  5501.000000   
mean       1.352692      1.636194      1.592516      1.832953     1.145246   
std        0.486944      0.494487      0.510654      0.424239     0.367534   
min        1.000000      1.000000      1.000000      1.000000     1.000000   
25%        1.000000      1.000000      1.000000      2.000000     1.000000   
50%        1.000000      2.000000      2.000000      2.000000     1.000000   
75%        2.000000      2.000000      2.000000      2.000000     1.000000   
max        4.000000      4.000000      4.000000      4.000000     4.000000   

               fin9         fin10  
count  15507.000000  15507.000000  
mean       1.248597      1.208938  
std        0.468721      0.441396  
min        1.000000      1.000000  
25%        1.000000      1.000000  
50%        1.000000      1.000000  
75%        1.000000      1.000000  
max        4.000000      4.000000

If you want a VERY quick stats overview of your data, use .describe()!

In [ ]:
# Display data types
data.dtypes

economy                 object
economycode             object
regionwb                object
pop_adult              float64
wpid_random              int64
                        ...   
remittances            float64
pay_onlne              float64
pay_onlne_mobintbuy    float64
pay_cash               float64
pay_cash_mobintbuy     float64
Length: 105, dtype: object

In [ ]:
# Display columns
data.columns

Index(['economy', 'economycode', 'regionwb', 'pop_adult', 'wpid_random', 'wgt',
       'female', 'age', 'educ', 'inc_q',
       ...
       'receive_wages', 'receive_transfers', 'receive_pension',
       'receive_agriculture', 'pay_utilities', 'remittances', 'pay_onlne',
       'pay_onlne_mobintbuy', 'pay_cash', 'pay_cash_mobintbuy'],
      dtype='object', length=105)

## Indexing and selecting the data

Usually, we only want to analyze a subset of the dataset. In this case, we select subset of rows and/or subset of columns.

In [7]:
# Filter to only include the economy column
data['economy']

0         Afghanistan
1         Afghanistan
2         Afghanistan
3         Afghanistan
4         Afghanistan
             ...     
154918       Zimbabwe
154919       Zimbabwe
154920       Zimbabwe
154921       Zimbabwe
154922       Zimbabwe
Name: economy, Length: 154923, dtype: object

Note that one column of a DataFrame is called a Series object!

In [ ]:
# Filter economy and economycode columns
data[['economy', 'economycode']]

economy economycode
0      Afghanistan         AFG
1      Afghanistan         AFG
2      Afghanistan         AFG
3      Afghanistan         AFG
4      Afghanistan         AFG
...            ...         ...
29777     Colombia         COL
29778     Colombia         COL
29779     Colombia         COL
29780     Colombia         COL
29781     Colombia         COL

[29782 rows x 2 columns]

In [ ]:
# Filter first two columns using iloc
data.iloc[:, :2]

economy economycode
0      Afghanistan         AFG
1      Afghanistan         AFG
2      Afghanistan         AFG
3      Afghanistan         AFG
4      Afghanistan         AFG
...            ...         ...
29777     Colombia         COL
29778     Colombia         COL
29779     Colombia         COL
29780     Colombia         COL
29781     Colombia         COL

[29782 rows x 2 columns]

**Note how we use the .iloc**. The first part of the bracket indicates the rows, the second part indicates the columns. 

Here's a couple of examples to guide you:

*   data.iloc[:, :5] - fetch ALL rows for the first 5 columns 
*   data.iloc[:2, :] - fetch the first 2 rows for ALL columns
*   data.iloc[1:3, 5:7] - fetch the 1st row until the 2nd row for 5th to 6th columns





In [ ]:
# Select all rows for economy and economycode columns
data.loc[:, 'economy':'economycode']

economy economycode
0      Afghanistan         AFG
1      Afghanistan         AFG
2      Afghanistan         AFG
3      Afghanistan         AFG
4      Afghanistan         AFG
...            ...         ...
29777     Colombia         COL
29778     Colombia         COL
29779     Colombia         COL
29780     Colombia         COL
29781     Colombia         COL

[29782 rows x 2 columns]

Note that we can use .loc in filtering. In .loc, we can use object type filters in filtering our rows and columns. In this case, we were able to filter columns from economy to economycode. This is more readable than using the .iloc!

In [ ]:
# Filter the data to display Philippine data
data[
    data['economy']=='Philippines'
]

Empty DataFrame
Columns: [economy, economycode, regionwb, pop_adult, wpid_random, wgt, female, age, educ, inc_q, emp_in, fin2, fin3, fin4, fin5, fin6, fin7, fin8, fin9, fin10, fin11a, fin11b, fin11c, fin11d, fin11e, fin11f, fin11g, fin11h, fin14a, fin14b, fin14c, fin15, fin16, fin17a, fin17b, fin19, fin20, fin21, fin22a, fin22b, fin22c, fin24, fin25, fin26, fin27a, fin27b, fin27c1, fin27c2, fin28, fin29a, fin29b, fin29c1, fin29c2, fin30, fin31a, fin31b, fin31c, fin32, fin33, fin34a, fin34b, fin34c1, fin34c2, fin35, fin36, fin37, fin38, fin39a, fin39b, fin39c1, fin39c2, fin40, fin41, fin42, fin43a, fin43b, fin43c1, fin43c2, fin44, fin45, fin46, fin47a, fin47b, fin47c1, fin47c2, fin47c3, fin47c4, fin47c5, mobileowner, fin48, account_fin, account_mob, account, saved, borrowed, receive_wages, receive_transfers, receive_pension, receive_agriculture, pay_utilities, ...]
Index: []

[0 rows x 105 columns]

To do filtering in pandas, you need to input a condition inside the bracket.

In this case, the condition is **data['economy'] == 'Philippines'**.

In [ ]:
# Filter Philippine correspondents with age < 30
data[
    (data['economy'] == 'Philippines') & (data['age'] < 30)
]

Empty DataFrame
Columns: [economy, economycode, regionwb, pop_adult, wpid_random, wgt, female, age, educ, inc_q, emp_in, fin2, fin3, fin4, fin5, fin6, fin7, fin8, fin9, fin10, fin11a, fin11b, fin11c, fin11d, fin11e, fin11f, fin11g, fin11h, fin14a, fin14b, fin14c, fin15, fin16, fin17a, fin17b, fin19, fin20, fin21, fin22a, fin22b, fin22c, fin24, fin25, fin26, fin27a, fin27b, fin27c1, fin27c2, fin28, fin29a, fin29b, fin29c1, fin29c2, fin30, fin31a, fin31b, fin31c, fin32, fin33, fin34a, fin34b, fin34c1, fin34c2, fin35, fin36, fin37, fin38, fin39a, fin39b, fin39c1, fin39c2, fin40, fin41, fin42, fin43a, fin43b, fin43c1, fin43c2, fin44, fin45, fin46, fin47a, fin47b, fin47c1, fin47c2, fin47c3, fin47c4, fin47c5, mobileowner, fin48, account_fin, account_mob, account, saved, borrowed, receive_wages, receive_transfers, receive_pension, receive_agriculture, pay_utilities, ...]
Index: []

[0 rows x 105 columns]

Here, we used two conditions joined by &. The conditions are also enclosed.

In [ ]:
# Filter data from United Kingdom OR United States
data[
    (data['economy']=='United Kingdom') | (data['economy']=='United States')
]

Empty DataFrame
Columns: [economy, economycode, regionwb, pop_adult, wpid_random, wgt, female, age, educ, inc_q, emp_in, fin2, fin3, fin4, fin5, fin6, fin7, fin8, fin9, fin10, fin11a, fin11b, fin11c, fin11d, fin11e, fin11f, fin11g, fin11h, fin14a, fin14b, fin14c, fin15, fin16, fin17a, fin17b, fin19, fin20, fin21, fin22a, fin22b, fin22c, fin24, fin25, fin26, fin27a, fin27b, fin27c1, fin27c2, fin28, fin29a, fin29b, fin29c1, fin29c2, fin30, fin31a, fin31b, fin31c, fin32, fin33, fin34a, fin34b, fin34c1, fin34c2, fin35, fin36, fin37, fin38, fin39a, fin39b, fin39c1, fin39c2, fin40, fin41, fin42, fin43a, fin43b, fin43c1, fin43c2, fin44, fin45, fin46, fin47a, fin47b, fin47c1, fin47c2, fin47c3, fin47c4, fin47c5, mobileowner, fin48, account_fin, account_mob, account, saved, borrowed, receive_wages, receive_transfers, receive_pension, receive_agriculture, pay_utilities, ...]
Index: []

[0 rows x 105 columns]

In [ ]:
# Save the filtered dataframe in a variable
philippine_data = data[
    data['economy'] == 'Philippines'
]

In [ ]:
# Show the saved data
philippine_data

Empty DataFrame
Columns: [economy, economycode, regionwb, pop_adult, wpid_random, wgt, female, age, educ, inc_q, emp_in, fin2, fin3, fin4, fin5, fin6, fin7, fin8, fin9, fin10, fin11a, fin11b, fin11c, fin11d, fin11e, fin11f, fin11g, fin11h, fin14a, fin14b, fin14c, fin15, fin16, fin17a, fin17b, fin19, fin20, fin21, fin22a, fin22b, fin22c, fin24, fin25, fin26, fin27a, fin27b, fin27c1, fin27c2, fin28, fin29a, fin29b, fin29c1, fin29c2, fin30, fin31a, fin31b, fin31c, fin32, fin33, fin34a, fin34b, fin34c1, fin34c2, fin35, fin36, fin37, fin38, fin39a, fin39b, fin39c1, fin39c2, fin40, fin41, fin42, fin43a, fin43b, fin43c1, fin43c2, fin44, fin45, fin46, fin47a, fin47b, fin47c1, fin47c2, fin47c3, fin47c4, fin47c5, mobileowner, fin48, account_fin, account_mob, account, saved, borrowed, receive_wages, receive_transfers, receive_pension, receive_agriculture, pay_utilities, ...]
Index: []

[0 rows x 105 columns]

## Computing summary statistics

In pandas, we can compute summary statistics in our sliced dataframe.

In [ ]:
# Compute count of correspondents for those in the Philippines
len(philippine_data)

0

In [ ]:
# Compute average age of correspondents in the Philippines
philippine_data['age'].mean()

nan

In [ ]:
# Compute median age of correspondents in the Philippines
philippine_data['age'].median()

nan

In [ ]:
# Compute min age of correspondents in the Philippines
philippine_data['age'].min()

nan

In [ ]:
# Compute min age of correspondents in the Philippines
philippine_data['age'].max()

nan

In [ ]:
# Compute number of respondents per income group
philippine_data['inc_q'].value_counts()

Series([], Name: inc_q, dtype: int64)

According to the **data description**, **inc_q** is the **within-economy household income quintile** (a version of percentile wherein we divide 100% into 5 groups).

Here's the details: 1 - poorest 20%, 2 - second 20%, 3 - middle 20%, 4 - fourth 20%
and 5 - richest 20%.

Based on the data, we have highest number of respondents in the richest 20% income group. **How will this affect our analysis?**

In [ ]:
# Compute unique education level entries
philippine_data['educ'].nunique()

0

## Checking for errors

To check for data errors, we sometimes check for missing data and duplicated entries.

In [ ]:
# Check for missing data
philippine_data.isna().sum()

economy                0.0
economycode            0.0
regionwb               0.0
pop_adult              0.0
wpid_random            0.0
                      ... 
remittances            0.0
pay_onlne              0.0
pay_onlne_mobintbuy    0.0
pay_cash               0.0
pay_cash_mobintbuy     0.0
Length: 105, dtype: float64

In [ ]:
# If we want, we can drop rows with NULLS in ANY columns
philippine_data.dropna()

Empty DataFrame
Columns: [economy, economycode, regionwb, pop_adult, wpid_random, wgt, female, age, educ, inc_q, emp_in, fin2, fin3, fin4, fin5, fin6, fin7, fin8, fin9, fin10, fin11a, fin11b, fin11c, fin11d, fin11e, fin11f, fin11g, fin11h, fin14a, fin14b, fin14c, fin15, fin16, fin17a, fin17b, fin19, fin20, fin21, fin22a, fin22b, fin22c, fin24, fin25, fin26, fin27a, fin27b, fin27c1, fin27c2, fin28, fin29a, fin29b, fin29c1, fin29c2, fin30, fin31a, fin31b, fin31c, fin32, fin33, fin34a, fin34b, fin34c1, fin34c2, fin35, fin36, fin37, fin38, fin39a, fin39b, fin39c1, fin39c2, fin40, fin41, fin42, fin43a, fin43b, fin43c1, fin43c2, fin44, fin45, fin46, fin47a, fin47b, fin47c1, fin47c2, fin47c3, fin47c4, fin47c5, mobileowner, fin48, account_fin, account_mob, account, saved, borrowed, receive_wages, receive_transfers, receive_pension, receive_agriculture, pay_utilities, ...]
Index: []

[0 rows x 105 columns]

In [ ]:
# Or, we can only drop rows with NULLS in certain columns
philippine_data.dropna(subset=['pay_cash_mobintbuy'])

Empty DataFrame
Columns: [economy, economycode, regionwb, pop_adult, wpid_random, wgt, female, age, educ, inc_q, emp_in, fin2, fin3, fin4, fin5, fin6, fin7, fin8, fin9, fin10, fin11a, fin11b, fin11c, fin11d, fin11e, fin11f, fin11g, fin11h, fin14a, fin14b, fin14c, fin15, fin16, fin17a, fin17b, fin19, fin20, fin21, fin22a, fin22b, fin22c, fin24, fin25, fin26, fin27a, fin27b, fin27c1, fin27c2, fin28, fin29a, fin29b, fin29c1, fin29c2, fin30, fin31a, fin31b, fin31c, fin32, fin33, fin34a, fin34b, fin34c1, fin34c2, fin35, fin36, fin37, fin38, fin39a, fin39b, fin39c1, fin39c2, fin40, fin41, fin42, fin43a, fin43b, fin43c1, fin43c2, fin44, fin45, fin46, fin47a, fin47b, fin47c1, fin47c2, fin47c3, fin47c4, fin47c5, mobileowner, fin48, account_fin, account_mob, account, saved, borrowed, receive_wages, receive_transfers, receive_pension, receive_agriculture, pay_utilities, ...]
Index: []

[0 rows x 105 columns]

In [ ]:
# Check for duplicated data in wpid_random (a unique identifier)
philippine_data.duplicated(subset='wpid_random', keep=False).sum()

0

**What does it mean?**

The subset checks for duplicated entries in wpid_random column. In this case, keep=False. Say, we have 4 duplicated entries. The keep=False parameter implies that we will mark ALL of these entries as duplicated. Thus, if we apply the .sum() method, it would result to 4.

Since there is no duplicated data in wpid_random column, we are sure that EACH respondent is unique!

In [ ]:
# Save the data
philippine_data.to_csv(
    'philippine_data.csv',
    index=False
    )

To save the data, use .to_csv() or .to_excel() depending on what file format you need.

Here, index=False because I don't want to save the index as column in my .csv file.

## Sample exercises

In [ ]:
# Select all female respondents in the Philippines


In [ ]:
# Select all female respondents with age < 30 


In [ ]:
# Select all female respondents that were able to complete secondary OR tertiary education 
